### ISSUE TITLE
Multiple Objective Optimization Module

### ISSUE THREAD

**Issue by rbharath** [deepchem/deepchem#1914]
A common task for DeepChem users is to design a molecule that satisfies a number of different objectives. For example, a user might want to design a molecule that is within a given solubility range, binds tightly to a given target, and does not bind to an antitarget. This isn't straightforward to do since there are multiple objectives. This is an example of the broader class of [multiple objective optimization](https://en.wikipedia.org/wiki/Multi-objective_optimization). 

I propose that we add a new `dc.moo` module that provides a general framework for multiparameter optimization in DeepChem. The core of this module would be a new abstract class `MultipleObjectiveOptimizer` that provides an abstract superclass specifying an API for multiple objective optimization problems. The core method for this class would have the following API

```
class MultipleObjectiveOptimizer:
  """Abstract superclass for multiple objective optimization.

  This class contains no implementations but rather specifies the API that must be satisfied by 

  def optimize(objectives, constraints, dataset):
    """The optimize method provides a generic API to frame a multiple objective optimization task.

    Parameters
    ------------
    objectives: list
      Each element of objectives should be a function that that accepts as argument datapoints in `dataset`. The function should return either a binary True/False value or a continuous readout.
    constraints: list
      Each element of `constraints` must be a string that is one of the following values `[True, False, "max", "min"]`. The `constraints` list is processed in order with `objectives`. That is, `constraints[i]` is the constraint for the function `objectives[i]`. If `constraints[i]` is `True` or `False`, this means that `objectives[i]` must be a function that returns binary values and that a solution to the multiple objective optimization problem must take on the corresponding binary value. If `constraints[i]` is "max" or "min", then this means that `objectives[i]` must be real valued and the goal is to either maximize or minimize this objective.
  dataset: dc.data.Dataset (optional, default=None)
      If specified, the dataset to optimize the multiple objectives over. `dataset.X` should contain these elements. `dataset.y`, `dataset.w`, `dataset.ids` don't need to be specified. If `dataset` is specified, `generator` should be `None`.
  generator: `Generator` (optional, default=None)
    If specified, a generator which on each invocation returns a new datapoint. This argument should be specified instead of `dataset` when there is a very large set which is to be optimized over and you don't necessarily want to store the optimized set of disk.

  Returns
  --------
  solutions: list[Tuple]
    Each entry in this list is a pareto optimal point in `dataset`. If no such points exist, `solutions == []`. Each point is a tuple of `(datapoint, values)` where `values` is a list containing the value of `datapoint` on `obectives`.
  """
```
There are a few existing python packages for multiple objective optimization such as [platypus](https://platypus.readthedocs.io/en/latest/index.html). We can wrap `platypus` and other solvers under this API so we have access to the best available tools. However these packages still appear to be relatively immature so we might also have to implement our own mutiple objective solvers. 

One very simpler optimizer could just walk over `dataset` and look for pareto optimal points, where a point is pareto optimal if it's not possible to improve any one objective function without worsening another [wikipedia](https://en.wikipedia.org/wiki/Pareto_efficiency)). (If `generator` is specified, all points up to a given budget `N` could be walked). Note that there are often multiple pareto optimal points, which form a set called the pareto frontier, which is why a list of solutions rather than a single solution is returned.

I'd love any feedback folks have on this proposed design. My hope is that it can become a powerful tool for molecular design when combined with DeepChem modules for solubility, potency, etc. I can take the lead on implementing this feature, but before I get started I want to make sure the design for this module is sensible.
  
CC @peastman @vsomnath @nd-02110114 @seyonechithrananda @lilleswing

**Comment by peastman** [comment]
If I understand correctly, you aren't talking about optimizing models?  You're assuming the user has already trained, for example, one model to predict solubility and another model to predict binding, and this feature would use those models to search for useful molecules.

I wonder if this is outside scope for DeepChem?  This isn't really about deep learning.  The models might just as easily be very simple statistical models, very complicated physics based models, or anything else that can assign a score to an input.

**Comment by rbharath** [comment]
Yes, that's correct. This is closer to a task of model inference rather than a question of model optimization. 

That's a good point about scope. I think this would be within DeepChem's scope since it's addressing the problem of doing inference with models in practice. While the functions can indeed be non-deep architectures, I think in practice deep models are used in practice with other models and there aren't good tools available to combine different models into one (or a few) meaningful predictions.

I've run into this problem a couple times where I can get all the way to trained DeepChem models for different tasks but I can't get one actual prediction easily out of the models. I think multiple objective optimization isn't well supported by the machine learning community yet so I think it makes sense for us to fill in the gap.

**Comment by nd-02110114** [comment]
I think this feature is a really helpful for Chem/Bio community.

> There are a few existing python packages for multiple objective optimization such as platypus. We can wrap platypus and other solvers under this API so we have access to the best available tools. However these packages still appear to be relatively immature so we might also have to implement our own mutiple objective solvers.

I recommend Optuna compared with platypus. Optuna is also used for multi objective optimization. 
The reason why I recommend is that Optuna has a large community (many maintainers and many users) compared with platypus. In terms of solvers, Optuna has possibility to support more solvers in the future  (Currently, just NSGA-II algorithm)
